docs/motivation.md

```diff
@@ -0,0 +1,550 @@
+up each method into what we call an "atomic" trait and combining them with a
+conjunction. For example, the definition of `Number` above is simply syntactic sugar for
+```verona
+type Number = {
+    add(self: Self, other: Number): Number
+} & {
+    sub(self: Self, other: Number): Number
+}
+```
+
+#### Capabilities and conjunction
+In Verona, each concrete type is tagged with a *reference capability*. The reference
+capability decides which operations are permitted through and with a reference
+depending on the viewpoint. For example, a reference with capability `mut` allows us to
+read and mutate the object, while a capability `iso` is unique and opaque to all
+read/write operations.
+
+The conjunction type operator plays well together with capabilities, allowing us
+to express capability tags as part of the regular type system. For example, given a class
+`C` we can express the type of a reference to an object of class type `C` with
+reference capability `mut` as
+```verona
+type CMut = C & mut
+```
+
+By extension this means that polymorphism in Verona doesn't need special
+handling of capabilities in type parameters. A type `A[T]` parameterized on `T`
+can simply be instantiated with a type like `C & mut`. The method types of `A`
+are then able to express further restrictions on `T` that allow us to define
+operations for each instantiation.
+```verona
+type A[T] = {
+    withImm(v : T & imm) : A[T] & imm // immutable A[T] created with immutable T
+    withMut(v : T & mut) : A[T] & mut // mutable A[T] created with mutable T
+}
+```
+
+* TODO: compare with
+    - objective c: categories
+    - ruby: open classes
+    - C#: extension methods
+
+### Disjunction types
+Verona also includes disjunction types. Assuming two types `C` and `D` we can
+define the new type
+```verona
+type COrD = C | D
+```
+This describes the union of the two types.
+
+If we assume `C` and `D` above are classes, we can furthermore write pattern
+matches to distinguish between the two types dynamically.
+
+Since in Verona there is no subtyping between classes, we are able to check that
+a pattern match is exhaustive, and furthermore matching on concrete types will
+allow us to do implicit type filtering over match cases. E.g. imagine a case
+match like this
+```verona
+// C, D, E are classes
+
+f(x : C | D | E, g : (D | E) -> S) : S {
+    match x with {
+        (y : C) => { // y : C
+            ...
+        }
+        _ => { // x : D | E
+            g x
+        }
+    }
+}
+```
+We are able to implicitly filter the type of `x` in the default case.
+
+In a language like ML, the corresponding code would look like
+```ocaml
+type t = C of c
+       | D of d
+       | E of e
+
+
+f (x : t) (g : t -> s) =
+    match x with
+        C y => ...
+        _   => g x
+```
+Note that `g` cannot have a more refined type than `t -> s`.
+
+#### Type discrimination
+In ML we might want to construct a type such as
```

Comment:
Don't introduce ML when trying to explain Verona. Instead, discuss open and closed world types, without reference to other languages.

Reply:
Define first, and compare second.

Perhaps want language comparison docs for each similar language. 

X-lang for Y programmers docs - often negative signposting.. e.g., common misunderstandings.